### DISCUSSION TITLE
Initialization Tasks

### DISCUSSION THREAD

**Discussion by maxfirefly** [nasa/fprime#2333] [Q&A]
Is there an "F Prime" way of initializing components? When I use the component constructors to initialize data members with values from an external library, segfaults are thrown. My best guess is that the external library being used is compiled after the component is compiled and instantiated, thus, the component is trying to construct its fields with a value that does not exist yet. Here is a hypothetical example:

```
  MyComponent ::
    MyComponent(
        const char *const compName
    ) : MyComponentComponentBase(compName)
  {
    data_member = ExternalLibray::value;
  }

```
`value` is a constant defined in `ExternalLibrary`. If `ExternalLibrary` is compiled after `MyComponent` is compiled and instantiated, `data_member` is initialized to null and causes problems downstream.

My specific questions are:
1. Is there a way to ensure that `ExternalLibrary` is compiled before `MyComponent` (and components in general) are compiled and instantiated?
2. If not, should I somehow define a `MyComponent::Initialize()` function to be executed during system boot up, effectively initializing during run time?
3. Is there an F Prime (or some other convention) way to handle this that I haven't mentioned?

**Comment by timcanham** [comment]
@maxfirefly The constructors are typically run statically before `main()`, so it's hard to predict when they are run with respect to other static initializers. It can be dependent on link order, so we recommend that you only initialize internal component state. The `init()` function is run after `main()` so you can safely add your code there, although to take advantage of the F Prime code generation you shouldn't add arguments to `init()`. If you require arguments, the typical pattern is to define a public method that can be called from the topology initialization code. There are examples in the reference code [here](https://github.com/nasa/fprime/blob/devel/Ref/Top/RefTopology.cpp#L85).

**Reply by LeStarch** [reply]
Just as a note: fprime doesn't add the `init()` function to the implementation anymore.  This was to discourage use of the `init()` function as this often led to arguments being added to the function, breaking the ability to use it in the autocoded topology.

While you could override `init()`, the current pattern is to add a `setup()` or `configure()` function, which may accept whichever args you need.  You'd need to call this specifically in the topology.

The inability to guarantee order of constructors is a language limitation.  It is specifically not guaranteed.
